arbiter-core/src/middleware.rs

```diff
@@ -0,0 +1,678 @@
+    filter_receivers: Arc<tokio::sync::Mutex<HashMap<ethers::types::U256, FilterReceiver>>>,
+}
+
+#[async_trait::async_trait]
+impl JsonRpcClient for Connection {
+    type Error = ProviderError;
+
+    /// Processes a JSON-RPC request and returns the response.
+    /// Currently only handles the `eth_getFilterChanges` call since this is
+    /// used for polling events emitted from the [`Environment`].
+    async fn request<T: Serialize + Send + Sync, R: DeserializeOwned>(
+        &self,
+        method: &str,
+        params: T,
+    ) -> Result<R, ProviderError> {
+        match method {
+            "eth_getFilterChanges" => {
+                // Get the `Filter` ID from the params `T`
+                // First convert it into a JSON `Value`
+                let value = serde_json::to_value(&params)?;
+
+                // Take this value as an array then cast it to a string
+                let str = value.as_array().ok_or(ProviderError::CustomError(format!(
+                    "The params value passed to the `Connection` via a `request` was empty. 
+                    This is likely due to not specifying a specific `Filter` ID!"
+                )))?[0]
+                    .as_str().ok_or(ProviderError::CustomError(format!(
+                        "The params value passed to the `Connection` via a `request` could not be later cast to `str`!"
+                    )))?;
+
+                // Now get the `U256` ID via the string decoded from hex radix.
+                let id = ethers::types::U256::from_str_radix(str, 16)
+                    .map_err(|e| ProviderError::CustomError(
+                        format!("The `str` representation of the filter ID could not be cast into `U256` due to: {:?}!", 
+                        e)))?;
+
+                // Get the corresponding `filter_receiver` and await for logs to appear.
+                let mut filter_receivers = self.filter_receivers.lock().await;
+                let filter_receiver =
+                    filter_receivers
+                        .get_mut(&id)
+                        .ok_or(ProviderError::CustomError(format!(
+                            "The filter ID does not seem to match any that this client owns!"
+                        )))?;
+                let mut logs = vec![];
+                let filtered_params = FilteredParams::new(Some(filter_receiver.filter.clone()));
+                while let Ok(received_logs) = filter_receiver.receiver.recv() {
+                    let ethers_logs = revm_logs_to_ethers_logs(received_logs);
+                    for log in ethers_logs {
+                        if filtered_params.filter_address(&log)
+                            && filtered_params.filter_topics(&log)
+                        {
+                            logs.push(log);
+                        }
+                    }
+                    break;
+                }
+
+                // TODO: This can probably be avoided somehow
```

Comment:
Should we try to address this before we merge tomorrow? I am okay ice-boxing things like this till after the minor release. In some sense thats what the patches are for. In either case we should make an issue for this with more detail on what needs to be done

Reply:
No I don't want to mess with this now. Icebox.